File: sdk/src/messages.js
// Canonical signing messages.
//
// These strings ARE the authorization. The node reconstructs each message
// from the request fields and checks the signature against it, so a message
// built even slightly differently — a reordered field, a decimal amount where
// base units were meant, a boolean rendered "True" instead of "true" — is not
// a bug that shows up as a wrong number. It shows up as "signature
// verification failed", with nothing to indicate which of a dozen fields was
// wrong.
//
// So every builder here takes explicit base-unit amounts and renders them the
// one way the chain expects. `test/conformance.test.mjs` signs each of these
// with the `hikma-wallet` CLI and asserts the signatures are byte-identical,
// which is the only way to be sure this file has not drifted from the Rust.

import { encodeAmount } from "./units.js";

/// Domain prefix for every native signature (mirrors `src/consensus/pos.rs`).
export const MESSAGE_PREFIX = "\x19Hikmalayer Signed Message:\n";

/// Address prefix.
export const ADDRESS_PREFIX = "hkm";

const bool = (value) => (value ? "true" : "false");

/// Move native HKM.
export const transfer = ({ from, to, amount, nonce }) =>
  `hikmalayer-transfer:${from}:${to}:${encodeAmount(amount)}:${nonce}`;

/// Bond stake to become (or top up) a validator. Uniquely among these
/// messages, the VRF public key is part of what is signed — it binds the
/// leader-election identity to the stake.
export const stake = ({ address, amount, nonce, vrfPublicKey }) =>
  `hikmalayer-stake:${address}:${encodeAmount(amount)}:${nonce}:${vrfPublicKey}`;

/// Begin unbonding stake.
export const withdraw = ({ address, amount, nonce }) =>
  `hikmalayer-withdraw:${address}:${encodeAmount(amount)}:${nonce}`;

/// Lock a grant with a cliff and linear release.
export const vest = ({ from, to, amount, cliffBlocks, durationBlocks, nonce }) =>
  `hikmalayer-vest:${from}:${to}:${encodeAmount(amount)}:${cliffBlocks}:${durationBlocks}:${nonce}`;

/// Issue a native token (HTS).
export const tokenCreate = ({ symbol, name, decimals, initialSupply, nonce }) =>
  `hikmalayer-token-create:${symbol}:${name}:${decimals}:${encodeAmount(initialSupply)}:${nonce}`;

/// Move token units.
export const tokenTransfer = ({ tokenId, to, amount, nonce }) =>
  `hikmalayer-token-transfer:${tokenId}:${to}:${encodeAmount(amount)}:${nonce}`;

/// Destroy the sender's token units, reducing supply.
export const tokenBurn = ({ tokenId, amount, nonce }) =>
  `hikmalayer-token-burn:${tokenId}:${encodeAmount(amount)}:${nonce}`;

/// Deposit into an HKM/token pool. `minShares` is the slippage bound.
export const ammAdd = ({ tokenId, amountHkm, amountToken, minShares, nonce }) =>
  `hikmalayer-amm-add:${tokenId}:${encodeAmount(amountHkm)}:${encodeAmount(amountToken)}` +
  `:${encodeAmount(minShares)}:${nonce}`;

/// Burn LP shares and withdraw the underlying assets.
export const ammRemove = ({ tokenId, shares, minHkm, minToken, nonce }) =>
  `hikmalayer-amm-remove:${tokenId}:${encodeAmount(shares)}:${encodeAmount(minHkm)}` +
  `:${encodeAmount(minToken)}:${nonce}`;

/// Trade against a pool. `hkmToToken` selects the direction.
export const ammSwap = ({ tokenId, hkmToToken, amountIn, minOut, nonce }) =>
  `hikmalayer-amm-swap:${tokenId}:${bool(hkmToToken)}:${encodeAmount(amountIn)}` +
  `:${encodeAmount(minOut)}:${nonce}`;

/// Write or revoke a credential.
export const credential = ({ id, subject, dataHash, revoke, nonce }) =>
  `hikmalayer-credential:${id}:${subject}:${dataHash}:${bool(revoke)}:${nonce}`;

export const messages = {
  transfer,
  stake,
  withdraw,
  vest,
  tokenCreate,
  tokenTransfer,
  tokenBurn,
  ammAdd,
  ammRemove,
  ammSwap,
  credential,
};

File: python-sdk/hikmalayer/messages.py
"""
Canonical signing messages for every Hikmalayer transaction type.

Every signature is bound to a network (chain_id), so a transaction signed
against the devnet can never be replayed on testnet or mainnet. The chain
prefixes the domain with the network id before hashing:

    <chain_id>:<domain>

These functions must match src/blockchain/transaction.rs exactly. The
conformance tests in the JS SDK verify parity against the hikma-wallet CLI;
this module mirrors those same domains.
"""

from typing import Optional


def _amt(value) -> str:
    """Amounts are always the exact decimal text of an integer."""
    return str(int(value))


def transfer(from_addr: str, to_addr: str, amount, nonce: int) -> str:
    return f"hikmalayer-transfer:{from_addr}:{to_addr}:{_amt(amount)}:{int(nonce)}"


def stake(address: str, amount, nonce: int, vrf_public_key: Optional[str] = None) -> str:
    base = f"hikmalayer-stake:{address}:{_amt(amount)}:{int(nonce)}"
    return f"{base}:{vrf_public_key}" if vrf_public_key else base


def withdraw(address: str, amount, nonce: int) -> str:
    return f"hikmalayer-withdraw:{address}:{_amt(amount)}:{int(nonce)}"


def vest(from_addr: str, to_addr: str, amount, cliff_blocks: int,
         duration_blocks: int, nonce: int) -> str:
    return (
        f"hikmalayer-vest:{from_addr}:{to_addr}:{_amt(amount)}:"
        f"{int(cliff_blocks)}:{int(duration_blocks)}:{int(nonce)}"
    )


def token_create(symbol: str, name: str, decimals: int,
                 initial_supply, nonce: int) -> str:
    return (
        f"hikmalayer-token-create:{symbol}:{name}:{int(decimals)}:"
        f"{_amt(initial_supply)}:{int(nonce)}"
    )


def token_transfer(token_id: str, to_addr: str, amount, nonce: int) -> str:
    # The sender is established by the key that signs this, so it does not
    # appear in the domain. Verified against `hikma-wallet sign-token-transfer`.
    return f"hikmalayer-token-transfer:{token_id}:{to_addr}:{_amt(amount)}:{int(nonce)}"


def token_burn(token_id: str, amount, nonce: int) -> str:
    # As above: no sender in the domain.
    return f"hikmalayer-token-burn:{token_id}:{_amt(amount)}:{int(nonce)}"


def amm_add(token_id: str, amount_hkm, amount_token,
            min_shares, nonce: int) -> str:
    return (
        f"hikmalayer-amm-add:{token_id}:{_amt(amount_hkm)}:{_amt(amount_token)}:"
        f"{_amt(min_shares)}:{int(nonce)}"
    )


def amm_remove(token_id: str, shares, min_hkm, min_token, nonce: int) -> str:
    return (
        f"hikmalayer-amm-remove:{token_id}:{_amt(shares)}:{_amt(min_hkm)}:"
        f"{_amt(min_token)}:{int(nonce)}"
    )


def amm_swap(token_id: str, hkm_to_token: bool, amount_in,
             min_out, nonce: int) -> str:
    direction = "true" if hkm_to_token else "false"
    return (
        f"hikmalayer-amm-swap:{token_id}:{direction}:{_amt(amount_in)}:"
        f"{_amt(min_out)}:{int(nonce)}"
    )


def credential(cred_id: str, subject: str, data_hash: str,
               revoke: bool, nonce: int) -> str:
    flag = "true" if revoke else "false"
    return f"hikmalayer-credential:{cred_id}:{subject}:{data_hash}:{flag}:{int(nonce)}"


def scoped(chain_id: str, domain: str) -> str:
    """
    Bind a domain to a network. Every signature the node accepts is over
    this scoped form, so a signature is only ever valid on one chain.
    """
    return f"{chain_id}:{domain}"
